probe-rs/src/probe/stlink/mod.rs

```diff
@@ -1909,8 +1890,18 @@ impl ArmMemoryInterface for StLinkMemoryInterface<'_> {
         })
     }
 
-    fn ap(&mut self) -> MemoryAp {
-        self.current_ap.clone()
+    fn get_parts(
+        &mut self,
+    ) -> Result<
+        (
+            &mut crate::architecture::arm::ArmCommunicationInterface<Initialized>,
+            &mut MemoryAp,
+        ),
+        DebugProbeError,
+    > {
+        Err(DebugProbeError::InterfaceNotAvailable {
```

Comment:
This is very confusing. Why would an ArmMemoryInterface implementation return "I don't speak ARM"?

Additionally, please consider following [C-CONV](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv) for the method names you introduce.

Reply:
The API requires it to be an `ArmCommunicationInterface` which the STLink probe is not.
If I understand correctly, this is because the STLink probe exposes a higher level API to access memory APs than the DAP interface. It’s still unclear to me why it is but I couldn’t get into that depth. (see the `get_arm_communication_interface` method.

I’ll be renaming this one as `try_as_parts`.

Reply:
I don’t like it either TBH. From what I could understand:
- we pack together the probe and the AP into a MemoryInterface that is passed to the sequences
- some times, some sequences need to do things directly with the probe and/or directly with the memory AP
- Any operation on a memory AP require a probe
- The operation on the memory AP cannot be abstracted away as they may not be present on all memory AP (I’m not fan of having loads of unimplemented/unsupported operation like what we see here)

I am yet to figure if those operations really require an ArmCommunicationInterface.
Also, the STLink probe, is the only one that’s not going through an `ArmCommunicationInterface`.